### ISSUE TITLE
OffsetInteger indexing seems to ignore offset

### ISSUE THREAD

**Issue by rdeits** [JuliaGeometry/GeometryTypes.jl#89]
I'm trying to translate some code to the new Face & OffsetInteger types, but I'm not sure that indexing is working properly. In the old v0.5 version of GeometryTypes, I could do:

```julia
julia> f = Face{3, Int, 1}(2, 2, 2)
Face(2,2,2)

julia> x = [1, 2, 3]
3-element Array{Int64,1}:
 1
 2
 3

julia> x[f]
(1,1,1)
```

I think the equivalent code in master would be:

```julia
julia> f = Face{3, OffsetInteger{1, Int}}(2, 2, 2)
3-element GeometryTypes.Face{3,GeometryTypes.OffsetInteger{1,Int64}}:
 |2-1|
 |2-1|
 |2-1|
```

but I get a different result:

```julia
julia> x[f]
(2, 2, 2)
```

Likewise, indexing with a single OffsetInteger also seems to ignore the offset:

```julia
julia> x = [1, 2, 3];

julia> x[OffsetInteger{1, Int}(2)]
2

julia> x[OffsetInteger{-1, Int}(2)]
2
```

or is that now how the OffsetInteger type works?

**Comment by rdeits** [comment]
More info: the MeshIO tests are currently failing for a possibly related reason. While testing the `.off` reader, MeshIO tries to construct a `GLTriangle(0, 1, 2)` which fails with an `InexactError()`, presumably because `GLTriangle` is a `Face{3, OffsetInteger{1, UInt32}}` and `OffsetInteger{1, UInt32}(0)` underflows.

**Comment by SimonDanisch** [comment]
Sorry for the confusion, the constructor now represents the number it got constructed with!
I'd have thought the printing helps here:
```Julia
julia> x = [GLTriangle(1, 1, 1)]
1-element Array{GeometryTypes.Face{3,GeometryTypes.OffsetInteger{1,UInt32}},1}:
 GeometryTypes.OffsetInteger{1,UInt32}[|1-1|, |1-1|, |1-1|]

julia> reinterpret(UInt32, x)
3-element Array{UInt32,1}:
 0x00000000
 0x00000000
 0x00000000
```
It was a bit tough to decide what the constructor should actually do. Let me make sure to discuss these things more openly in the future!
If you have good reasons to revert this and just pass through the actual number, let me know!

**Comment by SimonDanisch** [comment]
There is at least indirectly a test for this behaviour:
https://github.com/JuliaGeometry/GeometryTypes.jl/blob/master/test/faces.jl#L13

**Comment by rdeits** [comment]
Hm, I see. And was it intentional to flip the sign of the offset itself? Previously GLTriangle was a Face with offset = -1, but now it's a Face of OffsetIntegers with offset = +1. 

I guess I'm just not sure what the offset actually means in the new scheme. For example, when MeshIO loads a `.off` mesh, it reads in the three indices corresponding to a Face, which might be `"0 1 2"`. In the old code, I would write:

```julia
f = Face{3, Float64, -1}(parse.(Int, data))
```

while with the new code, I would write:

```julia
f = Face{3, Float64, OffsetInteger{1, Int}}(parse.(Int, data) .+ 1)
```

basically, I have to use the offset twice, once in the type parameter and once when I modify the data itself. Also, no matter what value I put for the offset, this will still index correctly, since

```julia
x[OffsetInteger{1, Int}(1)] === x[OffsetInteger{-1, Int}(1)]
```

Does this make more sense when communicating with OpenGL? I haven't thought about that side at all.

**Comment by SimonDanisch** [comment]
>I guess I'm just not sure what the offset actually means in the new scheme

It is now directly subtracted. Which is I guess my mental model of an offset in this context...
Since a constructor should represent the data it gets constructed from, I think this makes a lot of sense:
`x[OffsetInteger{OFF, Int}(IDX)] === x[IDX]`
What you're missing is something like reinterpret I guess.
For arrays of integers, I actually use exactly that. 
For construction, I just use that:
```Julia
GLTriangle(parse.(Int, 0based_data) + 1)
```
Which is not super nice, but pretty sure that shouldn't be a performance problem and it's also pretty concise. 
If you have an idea how to directly pass in the raw data without making this more verbose, let me know!
Btw, most decisions I try to justify here were also chosen to make the implementation simpler ;) The previous one was a bit awkward!

**Comment by rdeits** [comment]
Ok, thanks. I think I still prefer the previous behavior, but I will play around with MeshIO and see if I come around to your viewpoint ;-).

But I'm still confused about the change in the sign of the offset. The comment still describes it as an offset relative to Julia arrays, but the numerical offset for 0-based indexing has changed from -1 to +1. So is the offset now actually the offset of Julia relative to e.g. opengl instead?

**Comment by SimonDanisch** [comment]
I think the description might be the old one?
Feel free to change the sign :) If I recall correctly, I just followed my gut feeling there!

**Comment by rdeits** [comment]
Alright, I've played around with the new code with MeshIO and...I still don't like the new behavior :wink: 

To be clear, I think that moving the offset from the `Face` type to an `OffsetInteger` was a great idea, but I don't like the changes to the way the offset is actually handled. In particular, I think it's a bad sign that when dealing with zero-based data I end up:

1. starting with a zero-based index (let's say it's 0)
1. adding 1 to each index manually to convert to 1-based
1. calling `OffsetInteger{1, Int}(x)` which internally...
  1. subtracts 1 from x and stores the result (which is now 0 again)
1. using the result to index into an array, which internally....
  1. adds 1 back to x to get a one-based index

So we end up adding or subtracting 1 three separate times, which, even if it's trivial to the computer, is fairly confusing for my puny human brain. 

But rather than just whining, I figured I should come with an alternative, so here it is: https://github.com/rdeits/OffsetIndices.jl/blob/master/offsets.ipynb

Basically, I kept the constructor behavior of the old `GeometryTypes.Face`, so `ZeroIndexed(0)` gives something which will take the first element when used as an index. This version also has the nice property that the round trip from zero-based source to a zero-based sink never adds or subtracts 1 at all.  The big difference is that now it *matters* what index you use when constructing an `OffsetIndex` from a raw data, which I think is the right thing to do. 

What do you think? Am I making any sense?

**Comment by rdeits** [comment]
...and as soon as I started using my new OffsetIndex with MeshIO, I realized the problem with it (and the reason you chose the method you did). Since we allow `Face`s with regular Integer eltypes, we can end up in a situation where we need to do:

```julia
convert(Face{3, ZeroBased{Int}}, Face{3, Int}(1, 2, 3))
```

which presumably should give a face with the zero-based indices [0, 1, 2]. That implies that we should have `convert(ZeroBased{Int}, 1) == 1`. But I've simultaneously required that `ZeroBased(0)` should give you an index for the first element, so `ZeroBased(0) == 1`. We *could* have the constructor and `convert()` use different offsets, but that's *totally bonkers*. 

Instead, I'm moving farther into your camp with my next attempt: https://github.com/rdeits/OffsetIndices.jl/blob/master/offsets2.ipynb

In this attempt, as in your code, all raw integers are inherently one-based. I've flipped the sign of the offset, so it matches the offset in the old GeometryTypes, but otherwise this is just a slight cleanup of the code you already have. `ZeroIndex(1)` and `convert(ZeroIndex, 1)` now give the same answer. On the plus side, if we want really efficient mapping from zero-based data, `reinterpret()` does exactly the right thing:

```julia
reinterpret(OI.ZeroIndex{Int}, [0, 1, 2])
```

and I'm totally happy having `reinterpret()` and `convert()` do different things, since they're fundamentally different operations anyway. 

Sanity restored?
